x-pack/legacy/plugins/lens/public/indexpattern_plugin/dimension_panel/format_selector.tsx

```diff
@@ -0,0 +1,131 @@
+    }),
+  },
+  bytes: {
+    title: i18n.translate('xpack.lens.indexPattern.bytesFormatLabel', {
+      defaultMessage: 'Bytes (1024)',
+    }),
+  },
+};
+
+type FormatSelectorProps = Pick<IndexPatternDimensionPanelProps, 'data'> & {
+  selectedColumn: IndexPatternColumn;
+  currentIndexPattern: IndexPattern;
+  onChange: (newFormat?: { id: string; params?: Record<string, unknown> }) => void;
+};
+
+export function FormatSelector(props: FormatSelectorProps) {
+  const { selectedColumn, onChange } = props;
+
+  const currentFormat =
+    selectedColumn.params && 'format' in selectedColumn.params
+      ? selectedColumn.params.format
+      : undefined;
+
+  const selectedFormat = currentFormat?.id ? supportedFormats[currentFormat.id] : undefined;
+
+  const defaultOption = {
+    value: '',
+    label: i18n.translate('xpack.lens.indexPattern.defaultFormatLabel', {
+      defaultMessage: 'Default',
+    }),
+  };
+
+  const currentDecimals = currentFormat?.params?.decimals;
+
+  return (
+    <>
+      <EuiFormRow
+        label={i18n.translate('xpack.lens.indexPattern.columnFormatLabel', {
+          defaultMessage: 'Display format',
+        })}
+        display="rowCompressed"
+      >
+        <EuiComboBox
+          fullWidth
+          compressed
+          isClearable={false}
+          data-test-subj="indexPattern-dimension-format"
+          singleSelection={{ asPlainText: true }}
+          options={[defaultOption].concat(
+            Object.entries(supportedFormats).map(([id, format]) => ({
+              value: id || '',
+              label: format.title || id || '',
+            }))
+          )}
+          selectedOptions={
+            currentFormat
+              ? [
+                  {
+                    value: currentFormat.id || '',
+                    label: selectedFormat?.title || currentFormat.id || '',
+                  },
+                ]
+              : [defaultOption]
+          }
+          onChange={choices => {
+            if (choices.length === 0) {
+              return;
+            }
+
+            if (!choices[0].value) {
+              onChange();
+              return;
+            }
+            onChange({
+              id: choices[0].value,
+              params: { decimals: typeof currentDecimals === 'number' ? currentDecimals : 3 },
+            });
+          }}
+        />
+      </EuiFormRow>
+
+      {currentFormat ? (
+        <EuiFormRow
+          label={i18n.translate('xpack.lens.indexPattern.decimalPlacesLabel', {
+            defaultMessage: 'Decimals',
+          })}
+          display="rowCompressed"
+        >
+          <EuiFieldNumber
+            data-test-subj="indexPattern-dimension-formatDecimals"
+            value={typeof currentDecimals === 'number' ? currentDecimals : 3}
+            min={0}
+            max={20}
+            onChange={e => {
+              onChange({
+                id: (selectedColumn.params as { format: { id: string } })?.format!.id,
```

Comment:
No non-null assertion operator is needed here, since it cannot be null. (We'll hopefully introduce the linting rule for that soon, that has been released in a recent typescript-eslint release.)